doc/man/66-hpr.8.scd
66-hpr(8)

# NAME

66-hpr - Trigger the software shutdown procedure

# SYNOPSIS

66-hpr [ *-H* ] [ *-l* _live_ ] [ *-b* _banner_ ] [ *-f* ] [ *-h* | *-p* | *-r* ] [ *-d* | *-w* ] [ *-W* ]

# DESCRIPTION

*66-hpr* triggers the software shutdown procedure. Performs an immediate
hardware shutdown with the *-f* option.

It is normally invoked through *halt*, *poweroff* or *reboot* wrappers
installed by default at *%%skel%%*.

This program is a modified copy of *s6-linux-init-hpr* program.

- If the *-f* option is passed the system is stopped or rebooted immediately
  without properly cleaning up.
- Else, the machine's shutdown procedure is started.
- The command exits 0; the shutdown procedure happens asynchronously.

This is the traditional _sysvinit_ interface for the *halt*, *poweroff* and
*reboot* programs.

*66-hpr* must always be called with either *-h*, *-p* or *-r* options.

# OPTIONS

*-H*
	Prints this help.

*-l* _live_
	An absolute path. The directory to use. *66-hpr* sends signal to the
	listening pipe of the *66-shutdownd*(8) service at
	_live_/scandir/0/shutdownd/fifo. Default is *%%livedir%%*. The default can
	also be changed at compile-time by passing the --livedir=_live_ option to
	*./configure*. An absolute path is expected and should be under a writable
	filesystem - likely a RAM filesystem.

*-b*
	_banner_ . Text to display before executing the shutdown process. Default
	to:
```
*** WARNING *** 
The system is going down NOW!
```

*-f*
	Force. The command will not trigger a clean shutdown procedure; it will
	only sync the filesystems then tell the kernel to immediately halt,
	poweroff or reboot. This should be the last step in the lifetime of the
	machine.

*-h*
	Halt. The system will be shut down, but the power will remain connected.

*-p*
	Poweroff. Like halt but the power will also be turned off.

*-r*
	Reboot. The system will initialize a warm boot without disconnecting power.

*-d*
	Do not write a wtmp shutdown entry. See *utmp*(5).

*-w*
	Only write a wtmp shutdown entry; do not actually shut down the system.

*-W*
	Do not send a _wall_ message to users before shutting down the system. Some
	other implementations of the *halt*, *poweroff* and *reboot* commands use
	the *--no-wall* long option to achieve this.

# NOTES

*halt*, *poweroff* and *reboot* scripts that call *66-hpr* with the relevant
option *should be copied or symlinked* by the administrator into the binary directory of your system.


NEWS
Changelog for 66

In 0.2.5.2
----------

    - Fix build: Remove bytes.h oblibs header file
    
In 0.2.5.1
----------
    
    - Bugs fix
    - 66-tree -S options: if after_tree and tree have the same name,
    tree is considered as the very first tree to start.
    - 66-unmountall: do not umount SS_LIVE
    - Add SIGPWR control file at creation of .s6-svscan directory

In 0.2.5.0
----------
    
    - Bugs Fix: Bad memory allocation
    - 66-tree: add -S option which allow to order the start tree process
    even after it creation.
    
In 0.2.4.1
----------
    
    - Hot fix: fix html documentation

In 0.2.4.0
----------
    
    - Bugs fix: Bad memory allocation.
    
    - Enable again RB_DISABLE_CAD but don't crash if its not available
    and warn user
    
    - Add @optsdepends and @extdepends field at [main] section:
        - @optsdepends can be considere as: "enable one on this service 
        or none"
        - @extdepends can be considere as: "enable the service if it is 
        not already declared on a tree"
    
    - 66-in{tree,service}: 
        - in case of empty value the tools return None
        - add -n option: this avoids to display the name of the field.
        - add field at 66-intree:
            - start: displays the list of tree(s) started before
            - allowed: displays a list of allowed user to use the tree
            - symlinks: displays the target of tree's symlinks
        - add field at 66-inservice:
            - optsdepends: displays the service's optional dependencies
            - extdepends: displays the service's external dependencies
    
    - 66-shutdown skeleton: be safier and check if options are past
    
    - New tool:
        - 66-update: The 66-update program makes a complete transition 
        of trees and the live directory using a old 66 format 
        (the one being replaced) with the new 66 format.

In 0.2.3.2
----------
    
    - Fix bad annoucement at 66-svctl
    - Warn in case of bad key at parenthesis parsing process
    
In 0.2.3.1
----------
    
    - Bugs fix: bad allocation memory
    - Makefile: remove creation of empty directories. 66-tree will check
    all directories and create it if it missing.
    
In 0.2.3.0
----------
    
    - Adapt to oblibs 0.0.4.0
    - Remove RB_DISABLE_CAD at 66-boot by default and allow to enable
    it with -c option. Starting on virtual system like lxc crash with
    this option. Well be safe on every case.
    - Use of new function log_? familly from oblibs:
        - standardization of the exit code
        - standardization of the verbosity output
        - allow verbosity to 4 to display debug message
    - Doc typo fix        
    
In 0.2.2.2
----------
    
    - Adapt to skalibs 2.9.1.0
    - Remove insta_? deprecated function
    
In 0.2.2.1
----------
    
    - Typo fix at html documentation
    - Fix 66-intree without options
    
In 0.2.2.0
----------

    - Bugs fix: Always check the existing of the 66 heart directories
    - 66-info is now deprecated and splitted into two differents API.
    66-intree give informations about tree
    66-inservice give informations about services

In 0.2.1.2
----------

    - Hot fix: fix 66-info graph display.
    
In 0.2.1.1
----------

    - Hot fix: fix the build of the service dependencies graph.

In 0.2.1.0
----------

    - Bugs fix.
   
    - A synchronization is now made on reboot even with force option.
   
    - The field @name has no effects except for instantiated service but 
    can be omitted. The name of the frontend file is took as name. 
    In case of instantiated service, the field @name must contain the 
    complete name of the frontend service file -- Refer to the frontend 
    documentation page.
    
    - New skeleton file: rc.shutdown.final. This skeleton file will be run
    at the very end of the shutdown procedure, after all processes have 
    been killed and all filesystems have been unmounted, just before the 
    system is rebooted or the power turned off.
    This script normally remains empty.
   
    - Extra tools has removed and provided as an independent programs at
    https://framagit.org/Obarun/66-tools.git except for 66-echo and 
    66-unmountall program which are a dependent part of 66.
   
    - The account to run the s6-log program at the associate service logger
    can be set at compilation time by the --with-s6-log-user=USER options.
    The default is root. Obviously, the @runas field at the [logger] section
    overwritte it.
    This options set the account to run the uncaught-logs too. Also,
    this option can be overwritten with the '-l' option at 66-boot invocation
    and the '-L' option at 66-scandir invocation. 
    
In 0.2.0.4
----------
	
	- Bugs fix on memory allocation
	- Respect policies decision of user: 
		Remove -m option from 66-boot on init file
	- Add -d feature to 66-scandir 

In 0.2.0.3
----------

	- 66-parser: Fix write of configuration file, add -c|C features 
	
In 0.2.0.2
----------

	- Fix the written of user configuration file 
	- Fix oneshot and bundle status check
	- Add -l features to 66-tree, rewrite tree_unsupervise function
	
In 0.2.0.0
----------

	- New tools:
		- 66-boot.
		- 66-scanctl.
		- 66-shutdown.
		- 66-shutdownd.
		- 66-hpr.
		- 66-env.
		- 66-parser.
		- 66-which.
		- 66-echo.
		- 66-unmountall.
		- execl-subuidgid.
	- New @key field and change on frontend file:
		- [Logger] section accept a @depends field, @timestamp accept none
		as value, readiness notification is a default.
		- @hiercopy in [main] allow to copy any file or directory coming
		from the directory of the service.
	- 66-envfile is now deprecated, use execl-envfile in replacement.
	- 66-enable: add -F, -c, -C features.
	- 66-svctl is now asynchrone to launch services.
	- Man pages are available.
	- Environment variables limitation: maximum 100 files by directory,
	each file cannot contain more than 4096 bytes or 50 variables.

src/66/66-hpr.c
/* 
 * 66-hpr.c
 * 
 * Copyright (c) 2018-2020 Eric Vidal <eric@obarun.org>
 * 
 * All rights reserved.
 * 
 * This file is part of Obarun. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution.
 * This file may not be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file./
 *
 * This file is a modified copy of s6-linux-init-hpr.c file
 * coming from skarnet software at https://skarnet.org/software/s6-linux-init.
 * All credits goes to Laurent Bercot <ska-remove-this-if-you-are-not-a-bot@skarnet.org>
 * */

#include <skalibs/nonposix.h>

#include <unistd.h>
#include <signal.h>
#include <errno.h>
#include <utmpx.h>
#include <sys/reboot.h>

#include <oblibs/log.h>

#include <skalibs/sgetopt.h>
#include <skalibs/sig.h>
#include <skalibs/tai.h>
#include <skalibs/djbunix.h>
#include <skalibs/buffer.h>

#include <66/hpr.h>
#include <66/config.h>

#ifndef UT_NAMESIZE
#define UT_NAMESIZE 32
#endif

#ifndef UT_HOSTSIZE
#define UT_HOSTSIZE 256
#endif

#ifndef _PATH_WTMP
#define _PATH_WTMP "/dev/null/wtmp"
#endif

#define USAGE "66-hpr [ -H ] [ -l live ] [ -b banner ] [ -f ] [ -h | -p | -r ] [ -n ] [ -d | -w ] [ -W ]"

char const *banner = 0 ;
char const *live = 0 ;

static inline void info_help (void)
{
  static char const *help =
"66-hpr <options>\n"
"\n"
"options :\n"
"	-H: print this help\n" 
"	-l: live directory\n"
"	-b: end banner to display\n"
"	-f: force\n"
"	-h: halt the system\n"
"	-p: poweroff the system\n"
"	-r: reboot the system\n"
"	-d: do not write wtmp shutdown entry\n"
"	-w: only write wtmp shutdown entry\n"
"	-W: do not send a wall message\n"
;
	if (buffer_putsflush(buffer_1, help) < 0)
		log_dieusys(LOG_EXIT_SYS, "write to stdout") ;
}
int main (int argc, char const *const *argv)
{
	int what = 0 ;
	int force = 0 ;
	int dowtmp = 1 ;
	int dowall = 1 ;
	int dosync = 1 ;
	
	PROG = "66-hpr" ;
	{
		subgetopt_t l = SUBGETOPT_ZERO ;
		for (;;)
		{
			int opt = subgetopt_r(argc, argv, "Hl:hprfdwWb:n", &l) ;
			if (opt == -1) break ;
			switch (opt)
			{
				case 'H' : info_help() ; return 0 ;
				case 'l' : live = l.arg ; break ;
				case 'h' : what = 1 ; break ;
				case 'p' : what = 2 ; break ;
				case 'r' : what = 3 ; break ;
				case 'f' : force = 1 ; break ;
				case 'd' : dowtmp = 0 ; break ;
				case 'w' : dowtmp = 2 ; break ;
				case 'W' : dowall = 0 ; break ;
				case 'n' : dosync = 0 ; break ;
				case 'b' : banner = l.arg ; break ;
				default :  log_usage(USAGE) ;
			}
		}
		argc -= l.ind ; argv += l.ind ;
	}
	if (!banner) banner = HPR_WALL_BANNER ;
	if (live && live[0] != '/') log_die(LOG_EXIT_USER,"live: ",live," must be an absolute path") ;
	else live = SS_LIVE ;
	if (!what)
		log_die(LOG_EXIT_USER, "one of the -h, -p or -r options must be given") ;

	if (geteuid())
	{
		errno = EPERM ;
		log_diesys(LOG_EXIT_USER, "nice try, peon") ;
	}

	if (force)
	{
		if (dosync) sync() ;
		reboot(what == 3 ? RB_AUTOBOOT : what == 2 ? RB_POWER_OFF : RB_HALT_SYSTEM) ;
			log_dieusys(LOG_EXIT_SYS, "reboot()") ;
	}
	
	if (!tain_now_g()) log_warnsys("get current time") ;
	if (dowtmp)
	{
		struct utmpx utx =
		{
			.ut_type = RUN_LVL,
			.ut_pid = getpid(),
			.ut_line = "~",
			.ut_id = "",
			.ut_session = getsid(0)
		} ;
		strncpy(utx.ut_user, what == 3 ? "reboot" : "shutdown", UT_NAMESIZE) ;
		if (gethostname(utx.ut_host, UT_HOSTSIZE) < 0)
		{
			utx.ut_host[0] = 0 ;
			log_warnusys("gethostname") ;
		}
		else utx.ut_host[UT_HOSTSIZE - 1] = 0 ;

/* glibc multilib can go fuck itself */
#ifdef  __WORDSIZE_TIME64_COMPAT32
	{
		struct timeval tv ;
		if (!timeval_from_tain(&tv, &STAMP))
			log_warnusys("timeval_from_tain") ;
		utx.ut_tv.tv_sec = tv.tv_sec ;
		utx.ut_tv.tv_usec = tv.tv_usec ;
	}
#else
	if (!timeval_from_tain(&utx.ut_tv, &STAMP))
		log_warnusys("timeval_from_tain") ;
#endif

		updwtmpx(_PATH_WTMP, &utx) ;
	}
	if (dowall) hpr_wall(banner) ;
	if (dowtmp < 2)
	{
		size_t livelen = strlen(live) ;
		char tlive[livelen + INITCTL_LEN + 1] ;
		memcpy(tlive,live,livelen) ;
		memcpy(tlive + livelen,INITCTL,INITCTL_LEN) ;
		tlive[livelen + INITCTL_LEN] = 0 ;
		if (!hpr_shutdown(tlive,what, &tain_zero, 0))
			log_dieusys(LOG_EXIT_SYS, "notify 66-shutdownd") ;
	}
	return 0 ;
}
